ledger/simulation/resources.go

```diff
@@ -605,3 +607,696 @@ func (p *resourcePolicy) AvailableBox(app basics.AppIndex, name string, operatio
+			err = p.txnResources[i].addAccount(addr)
+			if err == nil {
+				return nil
+			}
+		}
+	}
+
+	// Finally try to find txn with room for both account and holding
+	for _, i := range p.appCallIndexes {
+		err = p.txnResources[i].addAssetHolding(addr, aid)
+		if err == nil {
+			return nil
+		}
+	}
+
+	return err
+}
+
+func (p *resourcePopulator) addLocal(addr basics.Address, aid basics.AppIndex) error {
+	var err error
+
+	// First try to find txn with account already available
+	for _, i := range p.appCallIndexes {
+		if p.txnResources[i].hasAccount(addr) {
+			err = p.txnResources[i].addApp(aid)
+			if err == nil {
+				return nil
+			}
+		}
+	}
+
+	// Then try to find txn with app already available
+	for _, i := range p.appCallIndexes {
+		if p.txnResources[i].hasApp(aid) {
+			err = p.txnResources[i].addAccount(addr)
+			if err == nil {
+				return nil
+			}
+		}
+	}
+
+	// Finally try to find txn with room for both account and app
+	for _, i := range p.appCallIndexes {
+		err = p.txnResources[i].addAppLocal(aid, addr)
+		if err == nil {
+			return nil
+		}
+	}
+
+	return err
+}
+
+func (p *resourcePopulator) populateResources(groupResourceTracker ResourceTracker, txnResources []ResourceTracker) error {
+	// We don't want to mutate the groupResourceTracker because it is used later in simulate for UnnamedResourcesAccessed
+	groupResources := struct {
+		Assets        []basics.AssetIndex
+		Apps          []basics.AppIndex
+		Accounts      []basics.Address
+		Boxes         []logic.BoxRef
+		AssetHoldings []ledgercore.AccountAsset
+		AppLocals     []ledgercore.AccountApp
+	}{
+		Apps:          []basics.AppIndex{},
+		Accounts:      []basics.Address{},
+		Boxes:         []logic.BoxRef{},
+		AssetHoldings: []ledgercore.AccountAsset{},
+		AppLocals:     []ledgercore.AccountApp{},
+	}
+
+	// Sort assets
+	for asset := range groupResourceTracker.Assets {
+		groupResources.Assets = append(groupResources.Assets, asset)
+	}
+	slices.SortFunc(groupResources.Assets, func(a, b basics.AssetIndex) int {
+		return cmp.Compare(a, b)
+	})
+
+	// Sort apps
+	for app := range groupResourceTracker.Apps {
+		groupResources.Apps = append(groupResources.Apps, app)
+	}
+	slices.SortFunc(groupResources.Apps, func(a, b basics.AppIndex) int {
+		return cmp.Compare(a, b)
+	})
+
+	// Sort accounts
+	for account := range groupResourceTracker.Accounts {
+		groupResources.Accounts = append(groupResources.Accounts, account)
+	}
+	slices.SortFunc(groupResources.Accounts, func(a, b basics.Address) int {
+		return cmp.Compare(a.GetUserAddress(), b.GetUserAddress())
+	})
+
+	// Sort boxes
+	// To sort boxes, we turn the app into a string, concat with the name, and then sort
+	for box := range groupResourceTracker.Boxes {
+		groupResources.Boxes = append(groupResources.Boxes, box)
+	}
+	slices.SortFunc(groupResources.Boxes, func(a, b logic.BoxRef) int {
+		return cmp.Compare(fmt.Sprintf("%d:%s", a.App, a.Name), fmt.Sprintf("%d:%s", b.App, b.Name))
+	})
```

Comment:
```suggestion
	for box := range groupResourceTracker.Boxes {
		groupResources.Boxes = append(groupResources.Boxes, box)
	}
	slices.SortFunc(groupResources.Boxes, func(a, b logic.BoxRef) int {
		return cmp.Or(cmp.Compare(a.App, b.App), cmp.Compare(a.Name, b.Name))
	})
```

Same for the compares below. Avoid the Sprintf, and make sure you're doing what you expect when the string length of you app/asset IDs differ.